src/repository.py

```diff
@@ -58,6 +155,188 @@ def __init__(self, repository: Repo, github_repository: Repository) -> None:
         self._github_repo = github_repository
         self._configure_git_user()
 
+    @cached_property
+    def base_path(self) -> Path:
+        """Return the Path of the repository.
+
+        Returns:
+            Path of the repository.
+        """
+        return Path(self._git_repo.working_tree_dir or self._git_repo.common_dir)
+
+    @property
+    def metadata(self) -> Metadata:
+        """Return the Metadata object of the charm."""
+        return get_metadata(self.base_path)
+
+    @property
+    def has_docs_directory(self) -> bool:
+        """Return whether the repository has a docs directory."""
+        return has_docs_directory(self.base_path)
+
+    @property
+    def current_branch(self) -> str:
+        """Return the current branch."""
+        try:
+            return self._git_repo.active_branch.name
+        except TypeError:
+            tag = next(
+                (tag for tag in self._git_repo.tags if tag.commit == self._git_repo.head.commit),
+                None,
+            )
+            if tag:
+                return tag.name
+            return self.current_commit
+
+    @property
+    def current_commit(self) -> str:
+        """Return the current branch."""
+        return self._git_repo.head.commit.hexsha
+
+    @property
+    def branches(self) -> set[str]:
+        """Return all local branches."""
+        return {branch.name for branch in self._git_repo.heads}
+
+    @contextmanager
+    def with_branch(self, branch_name: str) -> Iterator["Client"]:
+        """Return a context for operating within the given branch.
+
+        At the end of the 'with' block, the branch is switched back to what it was initially.
+
+        Args:
+            branch_name: name of the branch
+
+        Yields:
+            Context to operate on the provided branch
+        """
+        current_branch = self.current_branch
+
+        try:
+            yield self.switch(branch_name)
+        finally:
+            self.switch(current_branch)
+
+    @property
+    def summary(self) -> DiffSummary:
+        """Return a summary of the differences against the most recent commit."""
+        self._git_repo.git.add(".")
+
+        return DiffSummary.from_raw_diff(
+            self._git_repo.index.diff(None)
+        ) + DiffSummary.from_raw_diff(self._git_repo.head.commit.diff())
```

Comment:
I'm inclined to say that provided the two arguments we provide to `diff` (default or None) we should be pretty safe. I went back to the implementation of diff and we should be pretty safe as the two exceptions are filtered out. Of course we could be overly conservative and wrap everything around a try-except block, but really I don't think this would ever happen looking at the code

Reply:
I was more looking at the `git.add` part, since it does IO there are risks of exceptions there

Reply:
I have converted the property to a method